ledger-service/http-json/src/main/scala/com/digitalasset/http/Main.scala

```diff
@@ -86,68 +120,160 @@ object Main {
-            System.exit(ErrorCodes.StartupError)
-        }
-      case _ =>
-    }
-
-    val serviceF: Future[HttpService.Error \/ ServerBinding] =
-      HttpService.start(
-        startSettings = config,
-        contractDao = contractDao,
+    def program(shutdownHookCalled: Deferred[IO, Unit]) = {
+      import cats.FlatMap.ops.toAllFlatMapOps
+      val waitTime = 5.seconds
+      config.metricsReporter.foreach(reporter =>
+        logger.info(s"Using the $reporter metrics reporter")
+      )
+      val metricsReporting = new MetricsReporting(
+        getClass.getName,
+        config.metricsReporter,
+        JDuration.ofNanos(config.metricsReportingInterval.toNanos),
       )
+      def createActor[F[_]](name: String)(implicit F: Monad[F]): Resource[F, ActorSystem] =
+        Resource.make(F.pure(ActorSystem(name)))(asys =>
+          for {
+            _ <- F.pure(logger.debug(s"Terminating actor system (waiting up to $waitTime)"))
+            terminateFuture <- F.pure(asys.terminate())
+            _ <- F.pure(Await.ready(terminateFuture, waitTime))
+            _ = logger.debug("Actor system terminated")
+          } yield ()
+        )
+      def createMaterializer[F[_]](asys: ActorSystem)(implicit
+          F: Monad[F]
+      ): Resource[F, Materializer] =
+        Resource.make(F.pure(Materializer(asys)))(mat =>
+          F.pure(mat).flatMap(mat => F.pure(mat.shutdown()))
+        )
+      val resources =
+        for {
+          asys <- createActor[IO]("http-json-ledger-api")
+          _ = logger.debug("Started actor system")
+          mat <- createMaterializer[IO](asys)
+          metrics <- metricsReporting.acquire[IO]()
+          _ = logger.debug("Acquired metrics")
+        } yield (asys, mat, metrics)
+      resources
+        .use { case (asys, mat, metrics) =>
+          implicit val asysInstance: ActorSystem = asys
+          implicit val matInstance: Materializer = mat
+          implicit val sequencePool: AkkaExecutionSequencerPool = new AkkaExecutionSequencerPool(
+            "clientPool"
+          )(asys)
+          implicit val ec: ExecutionContextExecutor = asys.dispatcher
+          implicit val metricsI: Metrics = metrics
+          implicit val contextShift: ContextShift[IO] = IO.contextShift(ec)
+          implicit val timer: Timer[IO] = IO.timer(ec)
+          val contractDao =
+            config.jdbcConfig.map(c => ContractDao(c.driver, c.url, c.user, c.password))
 
-    discard {
-      sys.addShutdownHook {
-        HttpService
-          .stop(serviceF)
-          .onComplete { fa =>
-            logFailure("Shutdown error", fa)
-            terminate()
-          }
-      }
-    }
+          (contractDao, config.jdbcConfig) match {
+            case (Some(dao), Some(c)) if c.createSchema =>
+              logger.info("Creating DB schema...")
+              import dao.{jdbcDriver, logHandler}
+              Try(dao.transact(ContractDao.initialize).unsafeRunSync()) match {
```

Comment:
Is `dao.transact` supposed to throw if it is returning an `IO`? 
On another note, can't you thread it in the `IO` including the error handling instead of `..unsafeRunSync()) match{ .. IO(ErrorCodes.Ok)` ?